kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/events/AsyncEvent.java

```diff
@@ -0,0 +1,318 @@
+    }
+    currentState = State.PROCESSING;
+  }
+
+  public void transitionToOutputReady() {
+    if (!currentState.equals(State.PROCESSING)) {
+      log.error("[{}] Attempted to mark an async event as ready for output processing but "
+                    + "the event was not in the PROCESSING state", currentState.name());
+      throw new IllegalStateException("Cannot transition to OUTPUT_READY from the state "
+                                          + currentState.name());
+    }
+    currentState = State.OUTPUT_READY;
+  }
+
+  public void transitionToFinalizing() {
+    if (!currentState.equals(State.OUTPUT_READY)) {
+      log.error("[{}] Attempted to mark an async event as finalizing but the event was not "
+                    + "in the OUTPUT_READY state", currentState.name());
+      throw new IllegalStateException("Cannot transition to FINALIZING from the state "
+                                          + currentState.name());
+    }
+    currentState = State.FINALIZING;
+  }
+
+  public void transitionToDone() {
+    if (!currentState.equals(State.FINALIZING)) {
+      log.error(
+          "[{}] Attempted to mark an async event as DONE but the event not "
+              + "in the FINALIZING state",
+          currentState.name());
+      throw new IllegalStateException(
+          "Cannot transition to DONE from the state " + currentState.name());
+    } else if (!(outputForwards.isEmpty() && outputWrites.isEmpty())) {
+      log.error(
+          "[{}] Attempted to mark an async event as complete without draining all output queues"
+              + "first. Remaining forwards={} and remaining writes={}",
+          currentState.name(), outputForwards.size(), outputWrites.size());
+      throw new IllegalStateException(
+          "Can't transition to DONE when there are still records in the output buffers");
+    }
+
+    currentState = State.DONE;
+  }
+
+  public ProcessorRecordContext recordContext() {
+    return recordContext;
+  }
+
+  public long streamTime() {
+    return streamTime;
+  }
+
+  public long systemTime() {
+    return systemTime;
+  }
+
+  public int partition() {
+    return partition;
+  }
+
+  @SuppressWarnings("unchecked")
+  public <KIn> KIn inputKey() {
+    return (KIn) inputRecordKey;
+  }
+
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    final AsyncEvent that = (AsyncEvent) o;
+
+    if (partition != that.partition) {
+      return false;
+    }
+    if (systemTime != that.systemTime) {
+      return false;
+    }
+    if (streamTime != that.streamTime) {
+      return false;
+    }
+    if (!inputRecordKey.equals(that.inputRecordKey)) {
+      return false;
+    }
+    return Objects.equals(recordContext, that.recordContext);
+  }
+
+  @Override
+  public int hashCode() {
+    int result = inputRecordKey.hashCode();
```

Comment:
can't we just use `Objects.hashCode(...)` instead of computing the aggregated result ourselves?

Reply:
It's because of the processor record context. See the `#processorRecordContextHashCode` javadocs